File: frontend/src/core/game.ts
/** 企鹅棋核心规则引擎（纯函数式，不碰 DOM；与 Go 后端引擎保持同一套规则） */
import { boardCoords, HEX_DIRS } from './hex';
import type { GameState, Move, Player, Tile } from './types';

/** 按人数返回企鹅数：2 人 4 只、3 人 3 只、4 人 2 只 */
export function penguinCount(playerCount: number): number {
  return playerCount === 2 ? 4 : playerCount === 3 ? 3 : 2;
}

/** 生成随机棋盘：30 格 1 鱼、20 格 2 鱼、10 格 3 鱼（官方配置） */
function shuffledFish(): number[] {
  const bag: number[] = [];
  for (let i = 0; i < 30; i++) bag.push(1);
  for (let i = 0; i < 20; i++) bag.push(2);
  for (let i = 0; i < 10; i++) bag.push(3);
  for (let i = bag.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [bag[i], bag[j]] = [bag[j], bag[i]];
  }
  return bag;
}

/** 创建新对局 */
export function createGame(names: string[]): GameState {
  const coords = boardCoords();
  const fish = shuffledFish();
  const tiles: Tile[] = coords.map((c, i) => ({
    q: c.q,
    r: c.r,
    fish: fish[i],
    gone: false,
    owner: -1,
  }));
  const players: Player[] = names.map((name, i) => ({
    id: i,
    name,
    score: 0,
    penguins: new Array(penguinCount(names.length)).fill(-1),
    stuck: false,
  }));
  return { tiles, players, phase: 'placing', turn: 0, winner: -2, ply: 0 };
}

/** 邻居格索引（未沉没且存在），方向序与 HEX_DIRS 一致 */
export function neighbors(state: GameState, idx: number): number[] {
  const t = state.tiles[idx];
  const res: number[] = [];
  for (const [dq, dr] of HEX_DIRS) {
    const n = state.tiles.findIndex((x) => x.q === t.q + dq && x.r === t.r + dr);
    res.push(n);
  }
  return res;
}

/** 放置阶段的合法格：未沉没、无企鹅、恰好 1 条鱼 */
export function legalPlacements(state: GameState): number[] {
  return state.tiles
    .map((t, i) => ({ t, i }))
    .filter(({ t }) => !t.gone && t.owner === -1 && t.fish === 1)
    .map(({ i }) => i);
}

/** 某格企鹅的合法移动目标：六方向直线滑行，遇洞 / 企鹅 / 边界停止 */
export function legalMovesFrom(state: GameState, idx: number): number[] {
  const start = state.tiles[idx];
  const res: number[] = [];
  for (const [dq, dr] of HEX_DIRS) {
    let q = start.q + dq;
    let r = start.r + dr;
    for (;;) {
      const n = state.tiles.findIndex((x) => x.q === q && x.r === r);
      if (n === -1) break; // 出界
      const tile = state.tiles[n];
      if (tile.gone || tile.owner !== -1) break; // 洞或企鹅阻挡
      res.push(n);
      q += dq;
      r += dr;
    }
  }
  return res;
}

/** 玩家当前是否还有任何可行动的企鹅 */
export function playerCanMove(state: GameState, player: number): boolean {
  const p = state.players[player];
  for (const idx of p.penguins) {
    if (idx >= 0 && legalMovesFrom(state, idx).length > 0) return true;
  }
  return false;
}

/** 淘汰玩家（官方规则）：全部企鹅被困 → 企鹅离场，带走脚下格子的鱼，格子沉没 */
function eliminatePlayer(state: GameState, player: number): void {
  const p = state.players[player];
  for (let i = 0; i < p.penguins.length; i++) {
    const idx = p.penguins[i];
    if (idx >= 0) {
      state.tiles[idx].owner = -1;
      state.tiles[idx].gone = true;
      p.score += state.tiles[idx].fish;
      p.penguins[i] = -1;
    }
  }
  p.stuck = true;
}

/** 推进行动权：出局玩家跳过；无人可动则终局 */
function advanceTurn(state: GameState): void {
  const n = state.players.length;
  for (let step = 1; step <= n; step++) {
    const next = (state.turn + step) % n;
    if (state.players[next].stuck) continue; // 已出局
    if (playerCanMove(state, next)) {
      state.turn = next;
      return;
    }
    eliminatePlayer(state, next);
  }
  // 无人可动 → 结束
  finishGame(state);
}

/** 终局结算 */
function finishGame(state: GameState): void {
  state.phase = 'finished';
  let best = -1;
  let bestScore = -1;
  let tie = false;
  for (const p of state.players) {
    if (p.score > bestScore) {
      bestScore = p.score;
      best = p.id;
      tie = false;
    } else if (p.score === bestScore) {
      tie = true;
    }
  }
  state.winner = tie ? -1 : best;
}

/** 放置阶段总步数 */
function totalPlacements(state: GameState): number {
  return state.players.reduce((sum, p) => sum + p.penguins.length, 0);
}

/** 执行放置（已验证合法性） */
function doPlace(state: GameState, player: number, tileIdx: number): void {
  const tile = state.tiles[tileIdx];
  tile.owner = player;
  const p = state.players[player];
  const slot = p.penguins.indexOf(-1);
  p.penguins[slot] = tileIdx;
  state.ply++;
  if (state.ply >= totalPlacements(state)) {
    state.phase = 'moving';
    // 移动阶段由第一个放置的玩家（玩家 0）先行动
    state.turn = 0;
    if (!playerCanMove(state, state.turn)) {
      eliminatePlayer(state, state.turn);
      advanceTurn(state);
    }
  } else {
    state.turn = (state.turn + 1) % state.players.length;
  }
}

/** 执行移动（已验证合法性） */
function doMove(state: GameState, player: number, from: number, to: number): void {
  const fromTile = state.tiles[from];
  const toTile = state.tiles[to];
  fromTile.owner = -1;
  fromTile.gone = true; // 离开的格子沉没
  toTile.owner = player;
  state.players[player].score += fromTile.fish; // 起点格的鱼归玩家
  const p = state.players[player];
  p.penguins[p.penguins.indexOf(from)] = to;
  state.ply++;
  // 自己走完即被困 → 立即出局（官方规则：企鹅离场并带走脚下的鱼）
  if (!playerCanMove(state, player)) eliminatePlayer(state, player);
  advanceTurn(state);
}

/** 尝试走子：合法则应用并返回 true（返回新状态，不修改原状态） */
export function applyMove(state: GameState, mv: Move): GameState | null {
  if (state.phase === 'finished') return null;
  if (mv.player !== state.turn) return null;
  const next: GameState = structuredClone(state);
  if (mv.kind === 'place') {
    if (next.phase !== 'placing') return null;
    const t = next.tiles[mv.to];
    if (!t || t.gone || t.owner !== -1 || t.fish !== 1) return null;
    doPlace(next, mv.player, mv.to);
  } else {
    if (next.phase !== 'moving') return null;
    const from = next.tiles[mv.from];
    if (!from || from.owner !== mv.player) return null;
    if (!legalMovesFrom(next, mv.from).includes(mv.to)) return null;
    doMove(next, mv.player, mv.from, mv.to);
  }
  return next;
}

/** 当前局面下某玩家的全部合法着法（AI 与提示用） */
export function allMoves(state: GameState, player: number): Move[] {
  const res: Move[] = [];
  if (state.phase === 'finished') return res;
  if (state.phase === 'placing') {
    if (player === state.turn) {
      for (const i of legalPlacements(state)) res.push({ kind: 'place', player, from: -1, to: i });
    }
    return res;
  }
  if (player !== state.turn) return res;
  for (const idx of state.players[player].penguins) {
    if (idx < 0) continue;
    for (const to of legalMovesFrom(state, idx)) res.push({ kind: 'move', player, from: idx, to });
  }
  return res;
}

File: server/internal/game/game.go
// Package game 实现企鹅棋权威规则引擎（与前端 TS 引擎保持同一套规则）。
package game

import (
	"errors"
	"math/rand"
)

// Phase 对局阶段
type Phase string

const (
	PhasePlacing  Phase = "placing"
	PhaseMoving   Phase = "moving"
	PhaseFinished Phase = "finished"
)

// Tile 单个冰面格
type Tile struct {
	Q     int  `json:"q"`
	R     int  `json:"r"`
	Fish  int  `json:"fish"`
	Gone  bool `json:"gone"`
	Owner int  `json:"owner"` // -1 无企鹅
}

// Player 玩家
type Player struct {
	ID       int64  `json:"id"`
	Name     string `json:"name"`
	UserID   int64  `json:"userId"` // 数据库用户 ID，0 = 游客
	Score    int    `json:"score"`
	Penguins []int  `json:"penguins"` // -1 未放置
	Stuck    bool   `json:"stuck"`
}

// GameState 对局状态（与前端类型一一对应）
type GameState struct {
	Tiles   []Tile   `json:"tiles"`
	Players []Player `json:"players"`
	Phase   Phase    `json:"phase"`
	Turn    int      `json:"turn"`
	Winner  int      `json:"winner"` // -1 平局；-2 未结束
	Ply     int      `json:"ply"`
}

// Move 一步棋
type Move struct {
	Kind   string `json:"kind"` // place | move
	Player int    `json:"player"`
	From   int    `json:"from"`
	To     int    `json:"to"`
}

// hexDirs 六个滑行方向（axial 坐标）
var hexDirs = [6][2]int{
	{1, 0}, {1, -1}, {0, -1}, {-1, 0}, {-1, 1}, {0, 1},
}

// 常见错误
var (
	ErrNotYourTurn = errors.New("还没轮到你")
	ErrGameOver    = errors.New("对局已结束")
	ErrIllegalMove = errors.New("不合法的走子")
	ErrBadPhase    = errors.New("当前阶段不能这样走")
)

// boardCoords 官方 60 格棋盘：8 行，偶数行 8 格、奇数行 7 格
func boardCoords() []struct{ Q, R int } {
	cells := make([]struct{ Q, R int }, 0, 60)
	for row := 0; row < 8; row++ {
		cols := 8
		if row%2 == 1 {
			cols = 7
		}
		for col := 0; col < cols; col++ {
			q := col - (row-(row&1))/2
			cells = append(cells, struct{ Q, R int }{q, row})
		}
	}
	return cells
}

// penguinCount 按人数返回企鹅数
func penguinCount(playerCount int) int {
	if playerCount == 2 {
		return 4
	}
	if playerCount == 3 {
		return 3
	}
	return 2
}

// NewGame 创建随机新对局（2 人局）
func NewGame(name0, name1 string) *GameState {
	coords := boardCoords()
	// 鱼：30 格 1 鱼、20 格 2 鱼、10 格 3 鱼
	bag := make([]int, 0, 60)
	for i := 0; i < 30; i++ {
		bag = append(bag, 1)
	}
	for i := 0; i < 20; i++ {
		bag = append(bag, 2)
	}
	for i := 0; i < 10; i++ {
		bag = append(bag, 3)
	}
	rand.Shuffle(len(bag), func(i, j int) { bag[i], bag[j] = bag[j], bag[i] })

	tiles := make([]Tile, 60)
	for i, c := range coords {
		tiles[i] = Tile{Q: c.Q, R: c.R, Fish: bag[i], Owner: -1}
	}
	n := penguinCount(2)
	peng0 := make([]int, n)
	peng1 := make([]int, n)
	for i := range peng0 {
		peng0[i] = -1
		peng1[i] = -1
	}
	return &GameState{
		Tiles: tiles,
		Players: []Player{
			{ID: 0, Name: name0, Penguins: peng0, Score: 0},
			{ID: 1, Name: name1, Penguins: peng1, Score: 0},
		},
		Phase:  PhasePlacing,
		Turn:   0,
		Winner: -2,
	}
}

// Clone 深拷贝（AI 搜索用）
func (s *GameState) Clone() *GameState {
	tiles := make([]Tile, len(s.Tiles))
	copy(tiles, s.Tiles)
	players := make([]Player, len(s.Players))
	for i, p := range s.Players {
		peng := make([]int, len(p.Penguins))
		copy(peng, p.Penguins)
		players[i] = p
		players[i].Penguins = peng
	}
	return &GameState{
		Tiles:   tiles,
		Players: players,
		Phase:   s.Phase,
		Turn:    s.Turn,
		Winner:  s.Winner,
		Ply:     s.Ply,
	}
}

// find 按 axial 坐标找格索引，-1 不存在
func (s *GameState) find(q, r int) int {
	for i := range s.Tiles {
		if s.Tiles[i].Q == q && s.Tiles[i].R == r {
			return i
		}
	}
	return -1
}

// LegalPlacements 放置阶段合法格：未沉没、无企鹅、恰好 1 鱼
func (s *GameState) LegalPlacements() []int {
	res := make([]int, 0, 32)
	for i := range s.Tiles {
		t := &s.Tiles[i]
		if !t.Gone && t.Owner == -1 && t.Fish == 1 {
			res = append(res, i)
		}
	}
	return res
}

// LegalMovesFrom 某格企鹅的滑行目标：六方向直线，遇洞 / 企鹅 / 边界停止
func (s *GameState) LegalMovesFrom(idx int) []int {
	if idx < 0 || idx >= len(s.Tiles) {
		return nil
	}
	t := &s.Tiles[idx]
	res := make([]int, 0, 8)
	for _, d := range hexDirs {
		q, r := t.Q+d[0], t.R+d[1]
		for {
			n := s.find(q, r)
			if n == -1 {
				break
			}
			nt := &s.Tiles[n]
			if nt.Gone || nt.Owner != -1 {
				break
			}
			res = append(res, n)
			q += d[0]
			r += d[1]
		}
	}
	return res
}

// PlayerCanMove 玩家是否还有可行动的企鹅
func (s *GameState) PlayerCanMove(player int) bool {
	for _, idx := range s.Players[player].Penguins {
		if idx >= 0 && len(s.LegalMovesFrom(idx)) > 0 {
			return true
		}
	}
	return false
}

// AllMoves 玩家当前全部合法着法
func (s *GameState) AllMoves(player int) []Move {
	res := make([]Move, 0, 16)
	if s.Phase == PhaseFinished || s.Turn != player {
		return res
	}
	if s.Phase == PhasePlacing {
		for _, i := range s.LegalPlacements() {
			res = append(res, Move{Kind: "place", Player: player, From: -1, To: i})
		}
		return res
	}
	for _, idx := range s.Players[player].Penguins {
		if idx < 0 {
			continue
		}
		for _, to := range s.LegalMovesFrom(idx) {
			res = append(res, Move{Kind: "move", Player: player, From: idx, To: to})
		}
	}
	return res
}

// finishGame 终局结算
func (s *GameState) finishGame() {
	s.Phase = PhaseFinished
	best, bestScore, tie := -1, -1, false
	for _, p := range s.Players {
		if p.Score > bestScore {
			bestScore, best, tie = p.Score, int(p.ID), false
		} else if p.Score == bestScore {
			tie = true
		}
	}
	if tie {
		s.Winner = -1
	} else {
		s.Winner = best
	}
}

// eliminatePlayer 淘汰玩家（官方规则）：全部企鹅被困 → 企鹅离场，带走脚下格子的鱼，格子沉没
func (s *GameState) eliminatePlayer(player int) {
	p := &s.Players[player]
	for i, idx := range p.Penguins {
		if idx >= 0 {
			t := &s.Tiles[idx]
			t.Owner = -1
			t.Gone = true
			p.Score += t.Fish
			p.Penguins[i] = -1
		}
	}
	p.Stuck = true
}

// advanceTurn 推进行动权：出局玩家跳过；无人可动则终局
func (s *GameState) advanceTurn() {
	n := len(s.Players)
	for step := 1; step <= n; step++ {
		next := (s.Turn + step) % n
		if s.Players[next].Stuck {
			continue // 已出局
		}
		if s.PlayerCanMove(next) {
			s.Turn = next
			return
		}
		s.eliminatePlayer(next)
	}
	s.finishGame()
}

// ApplyMove 校验并执行一步棋（返回新状态，原状态不变）
func (s *GameState) ApplyMove(mv *Move) (*GameState, error) {
	if s.Phase == PhaseFinished {
		return nil, ErrGameOver
	}
	if mv.Player != s.Turn {
		return nil, ErrNotYourTurn
	}
	next := s.Clone()
	if err := next.applyChecked(mv); err != nil {
		return nil, err
	}
	return next, nil
}

// ApplyUnchecked 跳过回合校验直接应用（仅用于已确认合法的走法，如 AI 搜索）
func (s *GameState) ApplyUnchecked(mv *Move) error {
	if s.Phase == PhaseFinished {
		return ErrGameOver
	}
	return s.applyChecked(mv)
}

// applyChecked 已通过回合校验后的规则校验与应用
func (s *GameState) applyChecked(mv *Move) error {
	if mv.Kind == "place" {
		if s.Phase != PhasePlacing {
			return ErrBadPhase
		}
		if mv.To < 0 || mv.To >= len(s.Tiles) {
			return ErrIllegalMove
		}
		t := &s.Tiles[mv.To]
		if t.Gone || t.Owner != -1 || t.Fish != 1 {
			return ErrIllegalMove
		}
		t.Owner = mv.Player
		p := &s.Players[mv.Player]
		for i, pos := range p.Penguins {
			if pos == -1 {
				p.Penguins[i] = mv.To
				break
			}
		}
		s.Ply++
		total := 0
		for _, pl := range s.Players {
			total += len(pl.Penguins)
		}
		if s.Ply >= total {
			s.Phase = PhaseMoving
			// 移动阶段由第一个放置的玩家（玩家 0）先行动
			s.Turn = 0
			if !s.PlayerCanMove(s.Turn) {
				s.eliminatePlayer(s.Turn)
				s.advanceTurn()
			}
		} else {
			s.Turn = (s.Turn + 1) % len(s.Players)
		}
		return nil
	}

	// move
	if s.Phase != PhaseMoving {
		return ErrBadPhase
	}
	if mv.From < 0 || mv.From >= len(s.Tiles) || mv.To < 0 || mv.To >= len(s.Tiles) {
		return ErrIllegalMove
	}
	from := &s.Tiles[mv.From]
	if from.Owner != mv.Player {
		return ErrIllegalMove
	}
	legal := false
	for _, to := range s.LegalMovesFrom(mv.From) {
		if to == mv.To {
			legal = true
			break
		}
	}
	if !legal {
		return ErrIllegalMove
	}
	from.Owner = -1
	from.Gone = true
	s.Tiles[mv.To].Owner = mv.Player
	s.Players[mv.Player].Score += from.Fish
	p := &s.Players[mv.Player]
	for i, pos := range p.Penguins {
		if pos == mv.From {
			p.Penguins[i] = mv.To
			break
		}
	}
	s.Ply++
	// 自己走完即被困 → 立即出局（官方规则：企鹅离场并带走脚下的鱼）
	if !s.PlayerCanMove(mv.Player) {
		s.eliminatePlayer(mv.Player)
	}
	s.advanceTurn()
	return nil
}
